### PR TITLE
Changes to flow the TraceWriter into the SecretManager

### PR DESCRIPTION
**Pull request by fabiocav**

### CODE REVIEW

File: src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs [hunk 1]
```diff
@@ -13,16 +14,15 @@ internal static void Initialize(ScriptSettingsManager settingsManager, Container
         {
             builder.RegisterInstance(settingsManager);
 
-            // register the resolver so that it is disposed when the container
-            // is disposed
-            var webHostResolver = new WebHostResolver(settingsManager);
-            builder.RegisterInstance(webHostResolver);
+            builder.RegisterType<WebHostResolver>().SingleInstance();
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs]: Some of these changes are outside of the scope of the feature this PR introduces. I'm leaving this here for feedback, but we can revert a lot of these without problems.

File: src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs [hunk 2]
```diff
@@ -13,16 +14,15 @@ internal static void Initialize(ScriptSettingsManager settingsManager, Container
...
 
             // these services are externally owned by the WebHostResolver, and will be disposed
             // when the resolver is disposed
-            builder.Register<ISecretManager>(ct => webHostResolver.GetSecretManager(settings)).ExternallyOwned();
-            builder.Register<WebScriptHostManager>(ct => webHostResolver.GetWebScriptHostManager(settings)).ExternallyOwned();
-            builder.Register<WebHookReceiverManager>(ct => webHostResolver.GetWebHookReceiverManager(settings)).ExternallyOwned();
+            builder.RegisterType<DefaultSecretManagerFactory>().As<ISecretManagerFactory>().SingleInstance();
+            builder.Register<TraceWriter>(ct => ct.ResolveOptional<WebScriptHostManager>()?.Instance?.TraceWriter ?? NullTraceWriter.Instance).ExternallyOwned();
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs]: Not happy with this yet, but this gives us what we need for now. Getting to where we want to be (the broader logging changes we have planned) would address this.

File: src/WebJobs.Script.WebHost/Controllers/KeysController.cs [hunk 3]
```diff
@@ -18,11 +21,13 @@ public class KeysController : ApiController
...
         private readonly WebScriptHostManager _scriptHostManager;
         private readonly ISecretManager _secretManager;
+        private readonly TraceWriter _traceWriter;
 
-        public KeysController(WebScriptHostManager scriptHostManager, ISecretManager secretManager)
+        public KeysController(WebScriptHostManager scriptHostManager, ISecretManager secretManager, TraceWriter traceWriter)
         {
             _scriptHostManager = scriptHostManager;
             _secretManager = secretManager;
+            _traceWriter = traceWriter.WithPrefix($"{ScriptConstants.TracePrefixSecretManagement} API");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: This is what the "Source" property of TraceEvent is intended for. If you look at some core SDK logs, you'll see that they poplulate Source with the logical component, allowing log filtering, etc.

File: src/WebJobs.Script.WebHost/Controllers/KeysController.cs [hunk 4]
```diff
@@ -110,6 +115,8 @@ private IHttpActionResult AddOrUpdateFunctionSecret(string keyName, string value
                 operationResult = _secretManager.AddOrUpdateFunctionSecret(keyName, value, functionName);
             }
 
+            _traceWriter.VerboseFormat(Resources.TraceKeysApiSecretChange, keyName, functionName ?? "host", operationResult.Result);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: What does this new extension method add over doing simple string interpolation?

File: src/WebJobs.Script.WebHost/Properties/Resources.resx [hunk 5]
```diff
@@ -259,4 +259,28 @@
...
+    <value>{0} secret '{1}' for '{2}' {3}.</value>
+  </data>
+  <data name="TraceFunctionSecretGeneration" xml:space="preserve">
+    <value>Function ('{0}') secrets do not exist. Generating secrets.</value>
+  </data>
+  <data name="TraceHostSecretGeneration" xml:space="preserve">
+    <value>Host secrets do not exist. Generating secrets.</value>
+  </data>
+  <data name="TraceKeysApiSecretChange" xml:space="preserve">
+    <value>Secret '{0}' for '{1}' {2}.</value>
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Properties/Resources.resx]: Might be able to consolidate with fomat expression as TraceAddOrUpdateFunctionSecret above?

File: src/WebJobs.Script.WebHost/WebHooks/WebHookReceiverManager.cs [hunk 6]
```diff
@@ -116,15 +116,8 @@ protected virtual void Dispose(bool disposing)
...
-                    if (_httpConfiguration != null)
-                    {
-                        _httpConfiguration.Dispose();
-                    }
-
-                    if (_secretManager != null)
-                    {
-                        (_secretManager as IDisposable).Dispose();
-                    }
+                    _httpConfiguration.Dispose();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebHooks/WebHookReceiverManager.cs]: intentionally removing the null check on this?

File: src/WebJobs.Script/ScriptConstants.cs [hunk 7]
```diff
@@ -17,6 +17,8 @@ public static class ScriptConstants
         public const string TracePropertyIsUserTraceKey = "MS_IsUserTrace";
         public const string TracePropertyIsSystemTraceKey = "MS_IsSystemTrace";
 
+        public const string TracePrefixSecretManagement = "SecretManagement";
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptConstants.cs]: nit: might make more sense as a resx value?

File: src/WebJobs.Script.WebHost/Controllers/KeysController.cs (hunk 3 above)
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: Good point! I see that we have overloads of the various tracing methods, but think it would still be helpful to have this method, renamed to `WithSource`, so it automatically sets the source avoiding the need to have that done from every tracing call. Thoughts?

File: src/WebJobs.Script.WebHost/Controllers/KeysController.cs (hunk 4 above)
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Controllers/KeysController.cs]: When working with resource strings, string interpolation isn't an option, so this just reduces the verbosity over having to call `string.Format` on those tracing calls.

File: src/WebJobs.Script.WebHost/Properties/Resources.resx (hunk 5 above)
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Properties/Resources.resx]: Are you referring to this and TraceSecretDeleted below?

File: src/WebJobs.Script.WebHost/WebHooks/WebHookReceiverManager.cs (hunk 6 above)
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebHooks/WebHookReceiverManager.cs]: No, not intentional... good catch. The intention was to refactor this to `_httpConfiguration?.Dispose();`

**Comment by fabiocav** [comment]
@mathewc updated this to incorporate feedback and change the tracing logic to use the event source.
